### PR TITLE
Gzip the propfind response

### PR DESCRIPTION
**Pull request by rullzer**
Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

See also #20270 for the AppFramework version of this.

### CODE REVIEW

**Comment by Deltachaos** [comment]
I think it would make more sense to directly implement brotli instead as the server side performance effect it much less.

**Comment by rullzer** [comment]
> I think it would make more sense to directly implement brotli instead as the server side performance effect it much less.

Well I guess we should make this handler more generic indeed.
But as far as I know brotli is not natively available in php yet

**Comment by Deltachaos** [comment]
There is a PHP Extension. https://github.com/kjdev/php-ext-brotli

Even if currently not widely adopted, we could check for availability and use if available.

**Comment by rullzer** [comment]
Sure :)
That is why this PR is only a POC I hacked together. We could probably make it a more generic middleware. Properly parse the header so we return what is expected.

Feel free to push some commits against this branch to move this forward :)

File: apps/dav/lib/Connector/Sabre/PropfindGzipPlugin.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	public function gzipResponse(Request $request, Response $response) {
+		$header = $request->getHeader('Accept-Encoding');
+
+		if ($header === null) {
+			return $response;
+		}
+
+		if (strpos($header, 'gzip') !== false) {
+			$body = $response->getBody();
+			if (is_string($body)) {
```
Comment (rullzer) [inline apps/dav/lib/Connector/Sabre/PropfindGzipPlugin.php:66]: I guess for a followup we'd maybe want to also support streams? But for now this works surprisingly well.

**Comment by rullzer** [comment]
compres

Esp on slower mobile connections this will save a bunch.

**Comment by rullzer** [comment]
Lets do this. I'd rather have this is as soon as possible so we notice if something is wrong with it during the 20 cycle.

**Comment by kesselb** [comment]
What about a small test? ;)

**Comment by rullzer** [comment]
> What about a small test? ;)

added

**Review by kesselb — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]

**Comment by MorrisJobke** [comment]
🚀

**Pull request merged**